feat(TrackerStructures): Добавлены структуры для дальнейшей работы с …

Mono1it opened the pull request:
…проектом

sprint 14

Mono1it committed:
feat(TrackerStructures): Добавлены структуры для дальнейшей работы с проектом
feat(MainScreenVC): Добавлен UIcollectionView

Mono1it committed:
feat(ScheduleVC): Написал UI и логику для расписания
feat(CreateHabbitVC): Написал UI без логики

Mono1it committed:
feat(TrackerCell): сделал UI ячейки

Mono1it committed:
feat(TrackerVC): Добавлена анимация кнопки "+" и часть её логики
feat(CreateHabbitVC): Добавлена блокировка UI для кнопки "Создать"

Mono1it committed:
Странный файл который появился после коммита

Mono1it commented:
Сложный спринт, прошу оставить больше обратной связи по логике работы приложения и по взаимодействию UI.

NaR0RoW commented on Tracker.xcodeproj/xcuserdata/Ilya.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist:
Папку `xcuserdata` стоит игнорировать, кладя в `.gitignore`. В ней хранится информация о состоянии проекта (какой файл открыт, какой таргет выбран и т.п.), брейкпоинты и схемы для текущего пользователя `Xcode`.
Про `gitignore` можно прочитать вот [тут](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files)

NaR0RoW commented on Tracker/MainScreen/CreateHabbitModalViewController.swift:
```diff
@@ -0,0 +1,292 @@
+    func habbitViewController(_ controller: CreateHabbitModalViewController, didCreate tracker: Tracker, inCategory category: String)
+}
+
+class CreateHabbitModalViewController: UIViewController, UITextFieldDelegate, ScheduleViewControllerDelegate {
```
Стоит добавить модификатор `final`, чтобы вызовы методов работали быстрее.
Но это уже со звездочкой, хотя на собеседованиях спрашивают. Можно почитать про [диспачеризацию тут](https://habr.com/ru/articles/673400/)

NaR0RoW replied:
Ошибка в правописании:
`Habbit` -> `Habit`

Лучше включить [настройку](https://sarunw.com/posts/spell-checking-in-xcode/), чтобы `Xcode` подсвечивал ошибки в правописании

NaR0RoW commented on Tracker/MainScreen/CreateHabbitModalViewController.swift:
```diff
@@ -0,0 +1,292 @@
+    lazy var categoryTitle: String = "Важное"
+    lazy var trackerName: String = "Трекер"
+    lazy var trackerColor: UIColor = .ypGreen
+    lazy var weekdaysForTracker: [WeekDay] = []
```
Для достижения [инкапсуляции](https://habr.com/ru/articles/87205/) лучше отметить эти поля `private`

NaR0RoW commented on Tracker/MainScreen/CreateHabbitModalViewController.swift:
```diff
@@ -0,0 +1,292 @@
+        textField.clearButtonMode = .whileEditing
+        updateCreateButton()
+        
+        limitLabel.isHidden = updatedText.count <= 38
+        return updatedText.count <= 38
```
Лучше избегать магических чисел (38) и использовать константы. Имя константы будет раскрывать предназначение числа

NaR0RoW commented on Tracker/MainScreen/CreateHabbitModalViewController.swift:
```diff
@@ -0,0 +1,292 @@
+    // MARK: - Configure Methods
+    
+    private func configureTracker() -> TrackerCategory {
+        return TrackerCategory(title: self.categoryTitle, trackers: [Tracker(id: UUID(), name: self.trackerName, color: self.trackerColor, emoji: "😪", schedule: self.weekdaysForTracker)])
```
В однострочных выражениях `return` можно опускать

NaR0RoW replied:
`self` везде лучше опускать, если явно того не требуется

NaR0RoW commented on Tracker/MainScreen/ScheduleViewController.swift:
```diff
@@ -0,0 +1,141 @@
+    func scheduleViewController(_ controller: ScheduleViewController, didSelectDays days: [WeekDay])
+}
+
+class ScheduleViewController: UIViewController {
```
Стоит отметить как `final`. Перепроверь пожалуйста остальные классы в коде. Если класс не наследуется, то лучше отметить его как `final` для ускорения работы компилятора

NaR0RoW commented on Tracker/MainScreen/ScheduleViewController.swift:
```diff
@@ -0,0 +1,141 @@
+        dismiss(animated: true)
+    }
+    
+    @objc private func switchChanged(_ sender: UISwitch) {
+            let day = WeekDay.allCases[sender.tag]
+            if sender.isOn {
+                selectedDays.insert(day)
+            } else {
+                selectedDays.remove(day)
+            }
+            print("Выбранные дни: \(selectedDays.map { $0.name })")
+        }
```
Форматирование кода немного поехало, можно выделить это место и воспользоваться автоматическим форматированием в Xcode `control + I`

NaR0RoW commented on Tracker/MainScreen/TrackerStructures.swift:
```diff
@@ -0,0 +1,46 @@
+        }
+    }
+}
+    struct TrackerCategory {
+        let title: String
+        var trackers: [Tracker]
+    }
+    
+    struct TrackerRecord {
+        let trackerId: UUID
+        let date: String
+    }
```
Тут тоже форматирование поехало

NaR0RoW commented on Tracker/MainScreen/TrackerStructures.swift:
```diff
@@ -0,0 +1,46 @@
+}
+    struct TrackerCategory {
+        let title: String
+        var trackers: [Tracker]
```
Лучше иммутабельные структуры делать, то есть использовать поля - константы, а не переменные. Сами поля можно менять с помощью методов, которые будут возвращать новую структуру с измененными полями

NaR0RoW commented on Tracker/MainScreen/TrackersViewController.swift:
```diff
@@ -0,0 +1,273 @@
+        datePicker.datePickerMode = .date
+        datePicker.minimumDate = Date()
+        let localID = Locale.preferredLanguages.first
+        datePicker.locale = Locale(identifier: localID!)
```
Лучше не использовать `force` операции

NaR0RoW commented:
Логика работы привычек реализована не совсем корректно:
1.  В календаре должна быть возможность выбрать предыдущие даты.
2. Привычка должна отображаться только для выбранной даты.
3. Если привычки на выбранный день нет, то должна отображаться заглушка. 
4. Счетчик должен показывать количество отметок по привычке за всю историю.
5. Сейчас можно отметить трекер для будущей даты (такой возможности быть не должно).

https://github.com/user-attachments/assets/2ced43aa-725c-45fc-9f62-91e2a8188914

Mono1it committed:
feat(TrackerVC): Исправлена логика работы приложения
fix(TrackerStructures): Сделал структуры иммутабельными и сделал методы для их изменения
feat(TrackerCell): Изменил работу делегата кнопки записи привычки
